pr_agent/algo/ai_handler.py

```diff
@@ -92,17 +105,53 @@ async def chat_completion(self, model: str, system: str, user: str, temperature:
                     f"Generating completion with {model}"
                     f"{(' from deployment ' + deployment_id) if deployment_id else ''}"
                 )
-            response = await acompletion(
-                model=model,
-                deployment_id=deployment_id,
-                messages=[
-                    {"role": "system", "content": system},
-                    {"role": "user", "content": user}
-                ],
-                temperature=temperature,
-                azure=self.azure,
-                force_timeout=get_settings().config.ai_timeout
-            )
+            
+            bito_response = None
+            response = None
+            if model.startswith("bito"):
+                global trying
+                trying += 1
+                logging.info(f"trying {trying}/{OPENAI_RETRIES} to get AI prediction...")
+                model_type = model.replace("bito_", "").upper()
+                tmp = tempfile.NamedTemporaryFile(mode='w', delete=False) 
+                context_file = tmp.name
+
+                out = tempfile.NamedTemporaryFile(mode='w+t', delete=False)
+                out_file = out.name
+
+                with tempfile.NamedTemporaryFile(mode='w', delete=False) as system_prompt:
+                    system_prompt.write(system)
+                    instruction_file = system_prompt.name
+                with tempfile.NamedTemporaryFile(mode='w', delete=False) as user_prompt:
+                    user_prompt.write(user)
+                    cmd_output_file = user_prompt.name
+                bito_cmd = f"bito -p '{instruction_file}' -c '{context_file} -k {self.bito_cli_key}'"
+                bito_cmd += f" -f '{cmd_output_file}'" # Pass user prompt in this file
+                bito_cmd += f" -m '{model_type}'"
+                bito_cmd += f" > {out_file}"
+                proc = await asyncio.create_subprocess_shell(
+                    bito_cmd, stdout=None, 
+                    stderr=subprocess.PIPE
+                )
+                _, stderr = await proc.communicate()
+                if stderr:
+                    logging.info(f"Error from bito cli: {stderr.decode('utf-8')}")
+                else:
+                    with open(out_file, 'r') as file:
+                        bito_response = file.read()    
+                # bito_response = stdout.decode('utf-8')
+            else:
+                response = await acompletion(
+                    model=model,
+                    deployment_id=deployment_id,
+                    messages=[
+                        {"role": "system", "content": system},
+                        {"role": "user", "content": user}
+                    ],
+                    temperature=temperature,
+                    azure=self.azure,
+                    force_timeout=get_settings().config.ai_timeout
```

Comment:
**Security Issue**: The implementation of subprocess calls with asyncio.create_subprocess_shell without proper sanitization of inputs may lead to arbitrary command execution. This is particularly risky because the command string is constructed using external input, which can be manipulated by an attacker. <br> **Fix**: Avoid using asyncio.create_subprocess_shell with dynamic inputs. Instead, use asyncio.create_subprocess_exec with a list of arguments. Ensure that user input is validated and sanitized before being used in any command. <br> **Code Suggestion**: 
 ```
 async def safe_subprocess_exec(cmd, *args):
     sanitized_args = [sanitize(arg) for arg in args]
     return await asyncio.create_subprocess_exec(cmd, *sanitized_args)

 async def sanitize(input):
     # Sanitization logic here
     return sanitized_input
 ```